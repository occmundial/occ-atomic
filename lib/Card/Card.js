import React from 'react';
import PropTypes from 'prop-types';

import useStyles from './styles';

/** Container component with a card style */
const Card = ({ children, rest, raised, raisable, noPadding, className, style, id }) => {
    const classes = useStyles();

    const getElevation = () => {
        let classNames;
        const elevations = [{rest}, {raised}];
        const filtered = filter(elevations);
        if (filtered.length) classNames = classes[Object.keys(filtered[0])[0]];
        else classNames = classes.flat;
        if (raisable && !raised)
            classNames += ` ${classes.raisable}`;
        return classNames;
    };

    const filter = array => {
        return array.filter((item) => {
            const key = Object.keys(item)[0];
            if (item[key]) return key;
        });
    };

    return (
        <div id={id} className={`${!noPadding ? classes.card : classes.cardNoPadding} ${getElevation()}${className ? ` ${className}` : ''}`} style={style}>
            {children}
        </div>
    );
};

Card.propTypes = {
    children: PropTypes.node,
    /** The rest property cancels the elevation of the card. */
    rest: PropTypes.bool,
    /** The raised property gives a higher elevation to the card. */
    raised: PropTypes.bool,
    /** The raisable property gives a higher elevation to the card on hover. */
    raisable: PropTypes.bool,
    /** The card has by default a padding. Use this property if you need to remove it. */
    noPadding: PropTypes.bool,
    id: PropTypes.string,
    className: PropTypes.string,
    style: PropTypes.object
};

export default Card;
