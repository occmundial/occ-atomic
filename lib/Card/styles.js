import colors from '../subatomic/colors';
import shadows from '../subatomic/shadows';
import spacing from '../subatomic/spacing';
import grid from '../subatomic/grid';

const styles = {
    card: {
        background: colors.white,
        borderRadius: 6,
        padding: spacing.small,
        position: 'relative',
        transition: '0.3s box-shadow, 0.3s transform',
        [`@media screen and (min-width:${grid.sm}px)`]: {
            padding: spacing.medium
        }
    },
    cardNoPadding: {
        extend: 'card',
        padding: 0,
        [`@media screen and (min-width:${grid.sm}px)`]: {
            padding: 0
        }
    },
    flat: {
        boxShadow: shadows.lvl1,
        zIndex: 'auto'
    },
    rest: {
        boxShadow: shadows.lvl0,
        zIndex: 'auto'
    },
    raised: {
        boxShadow: shadows.lvl4,
        zIndex: 4
    },
    raisable: {
        '&:hover': {
            boxShadow: shadows.lvl4,
            transform: `translateY(-${spacing.xTiny}px)`,
            zIndex: 4
        }
    }
};

export default styles;
