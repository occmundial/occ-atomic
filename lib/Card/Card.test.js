import React from 'react';
import { mount } from 'enzyme';
import { JssProvider, SheetsRegistry } from 'react-jss';
import classnames from 'classnames';
import Card from './Card';

let sheets, classes;

describe("Card", () => {

    beforeEach(() => sheets = new SheetsRegistry());
    afterEach(() => sheets = null);

    it('matches the snapshot', () => {
        const wrapper = mount(
            <Card />
        );
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('has no padding', () => {
        const wrapper = mount(
            <JssProvider registry={sheets}>
                <Card noPadding />
            </JssProvider>
        );
        classes = sheets.registry[0].classes;
        expect(wrapper.find('div').prop('className').indexOf(classes.cardNoPadding)).not.toBe(-1);
    });

    it('has the right level of shadow', () => {
        let wrapper = mount(
            <JssProvider registry={sheets}>
                <Card />
            </JssProvider>
        );
        classes = sheets.registry[0].classes;
        expect(wrapper.find('div').prop('className')).toEqual(classnames(classes.card, classes.flat));
        wrapper = mount(
            <JssProvider registry={sheets}>
                <Card rest />
            </JssProvider>
        );
        classes = sheets.registry[0].classes;
        expect(wrapper.find('div').prop('className')).toEqual(classnames(classes.card, classes.rest));
        wrapper = mount(
            <JssProvider registry={sheets}>
                <Card raised />
            </JssProvider>
        );
        classes = sheets.registry[0].classes;
        expect(wrapper.find('div').prop('className')).toEqual(classnames(classes.card, classes.raised));
    });

});
