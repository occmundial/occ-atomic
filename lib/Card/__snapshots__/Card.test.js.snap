// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Card
    classes={
      Object {
        "card": "card",
        "cardNoPadding": "cardNoPadding",
        "flat": "flat",
        "raisable": "raisable",
        "raised": "raised",
        "rest": "rest",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": undefined,
      "className": "card flat",
      "id": undefined,
      "style": undefined,
    },
    "ref": null,
    "rendered": null,
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": undefined,
        "className": "card flat",
        "id": undefined,
        "style": undefined,
      },
      "ref": null,
      "rendered": null,
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Card styles matches the snapshot 1`] = `
Object {
  "card": Object {
    "@media screen and (min-width:768px)": Object {
      "padding": 32,
    },
    "background": "#ffffff",
    "borderRadius": 6,
    "padding": 16,
    "position": "relative",
    "transition": "0.3s box-shadow, 0.3s transform",
  },
  "cardNoPadding": Object {
    "@media screen and (min-width:768px)": Object {
      "padding": 0,
    },
    "extend": "card",
    "padding": 0,
  },
  "flat": Object {
    "boxShadow": "0 1px 2px 0 rgba(0,0,0,0.12), 0 1px 6px 0 rgba(0,29,124,0.05)",
    "zIndex": "auto",
  },
  "raisable": Object {
    "&:hover": Object {
      "boxShadow": "0 12px 9px -5px rgba(0,0,0,0.12), 0 10px 32px 0 rgba(0,29,124,0.12)",
      "transform": "translateY(-4px)",
      "zIndex": 4,
    },
  },
  "raised": Object {
    "boxShadow": "0 12px 9px -5px rgba(0,0,0,0.12), 0 10px 32px 0 rgba(0,29,124,0.12)",
    "zIndex": 4,
  },
  "rest": Object {
    "boxShadow": "0 0 0 0 rgba(0,0,0,0.0), 0 0 0 0 rgba(0,29,124,0)",
    "zIndex": "auto",
  },
}
`;

exports[`CardJSS matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Jss(Card) />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "classes": Object {
        "card": "Card-card-0-1-1",
        "cardNoPadding": "Card-cardNoPadding-0-1-2",
        "flat": "Card-flat-0-1-3",
        "raisable": "Card-raisable-0-1-6",
        "raised": "Card-raised-0-1-5",
        "rest": "Card-rest-0-1-4",
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "classes": Object {
          "card": "Card-card-0-1-1",
          "cardNoPadding": "Card-cardNoPadding-0-1-2",
          "flat": "Card-flat-0-1-3",
          "raisable": "Card-raisable-0-1-6",
          "raised": "Card-raised-0-1-5",
          "rest": "Card-rest-0-1-4",
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
